src/DonationForms/resources/app/DonationFormApp.tsx

```diff
@@ -38,44 +42,106 @@ const initialState = {
  * @since 3.0.0
  */
-function App() {
+function App({form}: { form: DonationForm }) {
+
     if (form.goal.isAchieved) {
         return (
             <DonationFormErrorBoundary>
-                <GoalAchievedTemplate goalAchievedMessage={form.settings.goalAchievedMessage}/>
+                <GoalAchievedTemplate goalAchievedMessage={form.settings.goalAchievedMessage} />
             </DonationFormErrorBoundary>
         );
     }
 
     if (form.design?.isMultiStep) {
         return (
-            <DonationFormSettingsProvider value={formSettings}>
+            <DonationFormSettingsProvider value={{...form.settings, ...donationFormNodeSettings}}>
                 <DonationFormStateProvider initialState={initialState}>
-                    <MultiStepForm sections={form.nodes} showHeader={form.settings?.showHeader} />
+                    <MultiStepForm form={form} />
                 </DonationFormStateProvider>
             </DonationFormSettingsProvider>
         );
     }
 
     return (
-        <DonationFormSettingsProvider value={formSettings}>
+        <DonationFormSettingsProvider value={{...form.settings, ...donationFormNodeSettings}}>
             <DonationFormStateProvider initialState={initialState}>
-                {form.settings?.showHeader && <Header />}
+                {form.settings?.showHeader && <Header form={form} />}
                 <Form defaultValues={defaultValues} sections={form.nodes} validationSchema={schema} />
             </DonationFormStateProvider>
         </DonationFormSettingsProvider>
     );
 }
 
+/**
+ * @unreleased
+ */
+function AppPreview() {
+
+    const {
+        subscribeToGoal,
+        subscribeToColors,
+        subscribeToSettings,
+        subscribeToCss,
+        unsubscribeAll
+    } = useDonationFormPubSub();
+
+    const [formState, setFormState] = useState<DonationForm>(form);
+
+    useEffect(() => {
+        subscribeToSettings((settings) => {
+            setFormState(prevState => {
+                return {
+                    ...prevState,
+                    settings: {
+                        ...prevState.settings,
+                        ...settings
+                    }
+                }
+            })
+        })
+
+        subscribeToGoal((goal) => {
+            setFormState(prevState => {
+                return {
+                    ...prevState,
+                    goal: {
+                        ...prevState.goal,
+                        ...goal
+                    }
+                }
+            })
+        })
+
+        subscribeToColors((data) => {
+            if (data['primaryColor']) {
+                root.style.setProperty('--givewp-primary-color', data['primaryColor']);
+            }
+
+            if (data['secondaryColor']) {
+                root.style.setProperty('--givewp-secondary-color', data['secondaryColor']);
+            }
+        })
+
+        subscribeToCss(({customCss}) => {
+            style.innerHTML = customCss;
+        })
```

Comment:
@alaca Is there a more specific style node property than `innerHTML`? Is it possible for someone to put  a close tag in the `customCss` variable and then add arbitrary HTML?

Reply:
@JasonTheAdams  we can probably use `innerText` instead of `innerHTML`.

Reply:
@JasonTheAdams I refactored this to use `innerText`. That should strip all HTML tags, scripts, etc.

Reply:
@JasonTheAdams I wasn't happy with the way how we set CSS so I refactored it a bit https://github.com/impress-org/givewp/pull/7059/commits/dd119797739da0beca0137a4d1d82d38a18bfd03

This should be pretty secure :smiley: